### PR TITLE
Redesign the neural network macro

### PR DESCRIPTION
**Pull request by tsoj**
```nim
network DemoNet:
  layers h, w:
    cv1:        Conv2D(@[1, h, w], 20, (5, 5))
    mp1:        Maxpool2D(cv1.out_shape, (2,2), (0,0), (2,2))
    cv2:        Conv2D(mp1.out_shape, 50, (5, 5))
    mp2:        MaxPool2D(cv2.out_shape, (2,2), (0,0), (2,2))
    fl:         Flatten(mp2.out_shape)
    hidden:     Linear(fl.out_shape[0], 500)
    classifier: Linear(500, 10)
  forward x:
    x.cv1.relu.mp1.cv2.relu.mp2.fl.hidden.relu.classifier
```

creates now something like this
<details>
  <summary>Click to expand!</summary>

```nim
type
  DemoNet[T] = object
    cv1: Conv2D[T]
    mp1: Maxpool2D[T]
    cv2: Conv2D[T]
    mp2: MaxPool2D[T]
    fl: Flatten[T]
    hidden: Linear[T]
    classifier: Linear[T]

proc init[T](ctx: Context[AnyTensor[T]]; model_type: typedesc[DemoNet[T]]; h: auto; w: auto): DemoNet[T] =
  template cv1(): auto =
    result.cv1

  template mp1(): auto =
    result.mp1

  template cv2(): auto =
    result.cv2

  template mp2(): auto =
    result.mp2

  template fl(): auto =
    result.fl

  template hidden(): auto =
    result.hidden

  template classifier(): auto =
    result.classifier

  cv1 = init(ctx, Conv2D[T], @[1, h, w], 20, (5, 5))
  mp1 = init(ctx, Maxpool2D[T], cv1.out_shape, (2, 2), (0, 0), (2, 2))
  cv2 = init(ctx, Conv2D[T], mp1.out_shape, 50, (5, 5))
  mp2 = init(ctx, MaxPool2D[T], cv2.out_shape, (2, 2), (0, 0), (2, 2))
  fl = init(ctx, Flatten[T], mp2.out_shape)
  hidden = init(ctx, Linear[T], fl.out_shape[0], 500)
  classifier = init(ctx, Linear[T], 500, 10)

proc forward[T](self: DemoNet[T]; x: auto): auto =
  template cv1(x: varargs[untyped]): auto =
    forward(self.cv1, x)

  template mp1(x: varargs[untyped]): auto =
    forward(self.mp1, x)

  template cv2(x: varargs[untyped]): auto =
    forward(self.cv2, x)

  template mp2(x: varargs[untyped]): auto =
    forward(self.mp2, x)

  template fl(x: varargs[untyped]): auto =
    forward(self.fl, x)

  template hidden(x: varargs[untyped]): auto =
    forward(self.hidden, x)

  template classifier(x: varargs[untyped]): auto =
    forward(self.classifier, x)

  x.cv1.relu.mp1.cv2.relu.mp2.fl.hidden.relu.classifier
```
</details>

This would make adding new layers easier than it is now (where one would need to dig deep into macro territory), and it also allows custom layers that don't need to be included in the core macro.

For example, the `Linear` layer could be implemented like this:
```nim
type
  Linear*[T] = object
    weight*: Variable[Tensor[T]]
    bias*: Variable[Tensor[T]]

proc init*[T](
  ctx: Context[Tensor[T]],
  layer_type: typedesc[Linear[T]],
  num_input, num_output: int
): Linear[T] =
  result.weight = ctx.variable(kaiming_normal([num_output, num_input], T), requires_grad = true)
  result.bias = ctx.variable(zeros[T]([1, num_output]), requires_grad = true)

proc forward*[T](self: Linear[T], input: Variable[Tensor[T]]): Variable[Tensor[T]] =
  input.linear(weight = self.weight, bias = self.bias)

proc out_shape*[T](self: Linear[T]): seq[int] =
  @(self.weight.value.shape)
proc in_shape*[T](self: Linear[T]): seq[int] =
  @(self.weight.value.shape)
```

I tested this with `ex02_handwritten_digits_recognition.nim` and it seems to work fine (97.74% accuracy after 2 epochs).

Though this is only a prototype to get the discussion started, so there are a few questions still open:
- is this even wanted
- can the layertype-to-neuralmacro-interface be improved?
- test for performance regressions
- check if the layers are implemented correctly, especially GRU and GCN
- there could be identifier ambiguities in some cases, e.g. if I do something like this:
```nim
forward self:
    self.cv1.relu.mp1.cv2.relu.mp2.fl.hidden.relu.classifier
```
- maybe I should add better error message for the macro
- what about layers based on other tensor types (e.g. `CudaTensor`)?
- documentation

### CODE REVIEW

**Comment by tsoj** [comment]
I updated the other examples in [`examples/`](https://github.com/mratsim/Arraymancer/tree/master/examples) and they all seem to work fine. Though I haven't tested the GCN layer yet.

**Comment by mratsim** [comment]
That is awesome.

Yes it's wanted, basically one of the big limitation of the current macro is that you can compose models and this seems to solve that neatly.

Yes Cuda-based network needs the same but none exist yet.

**Comment by tsoj** [comment]
@anon767 I noticed that mentions in review threads don't seem to work, if they do sorry for spamming you :/

I am trying to redesign the neural network mini-language, such that adding new layers is easier to do. Unfortunately, I haven't used GCNs so far, so I pretty much just copy-pasted the stuff that you added in https://github.com/mratsim/Arraymancer/pull/503.

Just in case that you have time: Does [this implementation](https://github.com/mratsim/Arraymancer/blob/b242256ff22468ac56df0d0f0db49d556193a22f/src/arraymancer/nn/layers/gcn.nim#L107) of the GCN layer wrapper look correct to you?

If you have an application that used the Arraymancer neural net mini-lang with GCNs, it would be cool (in case that it's alright for you) if you could send it to me, so I can try if it still works with the new [network](https://github.com/mratsim/Arraymancer/blob/b242256ff22468ac56df0d0f0db49d556193a22f/src/arraymancer/nn_dsl.nim#L334) macro.

**Comment by tsoj** [comment]
@mratsim Is there anything I can do to make this easier to merge? (I understand that you might not have time for this currently.)

File: examples/ex02_handwritten_digits_recognition.nim [hunk 1]
```diff
@@ -33,26 +33,48 @@ let
...
-    hidden:     Linear(fl.out_shape, 500)
-    classifier: Linear(500, 10)
   forward x:
-    x.cv1.relu.mp1.cv2.relu.mp2.fl.hidden.relu.classifier
+    x.cv1.relu.mp1.cv2.relu.mp2.fl
+
+proc out_shape*[T](self: SomeConvNet[T]): seq[int] =
+  self.fl.out_shape
+proc in_shape*[T](self: SomeConvNet[T]): seq[int] =
+  self.cv1.in_shape
```
Comment (mratsim) [inline examples/ex02_handwritten_digits_recognition.nim:61]: A section about "Composing models" can be added in the README to mention this convention

File: examples/ex06_shakespeare_generator.nim [hunk 2]
```diff
@@ -124,121 +124,23 @@ proc sample[T](probs: Tensor[T]): int =
...
-  let encoded = model.encode(input)
-  let (output, hiddenN) = model.gru_forward(encoded, hidden0)
-
-  # result.output is of shape [Sequence, BatchSize, HiddenSize]
-  # In our case the sequence is 1 so we can simply flatten
-  let flattened = output.reshape(output.value.shape[1], HiddenSize)
-
-  result.output = model.decode(flattened)
-  result.hidden = hiddenN
-
```
Comment (mratsim) [inline examples/ex06_shakespeare_generator.nim:241]: Very nice

File: README.md [hunk 3]
```diff
@@ -327,23 +326,28 @@ echo answer
...
+    cv:         SomeConvNet(h, w)
+    hidden:     Linear(cv.fl.out_shape[0], 500)
+    classifier: Linear(hidden.out_shape[0], 10)
+  forward x:
+    x.cv.hidden.relu.classifier
+
+let
+  ctx = newContext Tensor[float32] # Autograd/neural network graph
+  model = ctx.init(DemoNet, 28, 28)
+  optim = model.optimizerSGD(learning_rate = 0.01'f32)
```
[the remark this answers is not in the document]
Comment (tsoj) [reply README.md:344]: makes sense, i changed it

**Comment by mratsim** [comment]
Awesome. I think we can release that and tag a new version just after with backward compat notes.

File: src/arraymancer/nn_dsl.nim [hunk 4]
```diff
@@ -2,11 +2,406 @@
...
+            ident"result",
+            layerInfo.name
+          )
+        )
+      )
+    )
+
+  let
+    ctxSymbol = genSym(nskParam, "ctx")
+    underlyingTypeSymbol = ident($toStrLit(genSym(nskGenericParam, "T")))# TODO fix this
```
Comment (Vindaar) [inline src/arraymancer/nn_dsl.nim:159]: Why the back and forth conversion from an `nnkSym` back to a `nnkIdent`? Does a symbol cause problems here?

File: src/arraymancer/nn_dsl.nim [hunk 5]
```diff
@@ -2,11 +2,406 @@
...
+
+  # 2. create model type
+  let modelType = createModelType(layerInfos, modelName)
+
+  # 3. create init proc
+  let initProc = createInitProc(layerInfos, sections.layers, modelName)
+
+  # 4. create forward proc
+
+  let forwardProc = createForwardProc(layerInfos, sections.forward, modelName)
```
Comment (Vindaar) [inline src/arraymancer/nn_dsl.nim:398]:
```suggestion
  # 4. create forward proc
  let forwardProc = createForwardProc(layerInfos, sections.forward, modelName)
```

**Comment by Vindaar** [comment]
Would you be willing to change this page in the docs:
https://mratsim.github.io/Arraymancer/howto.perceptron.html
and as source:
https://github.com/mratsim/Arraymancer/blob/master/docs/howto.perceptron.rst
as I'm not quite sure yet what actually changes in that example.

The docs themselves also have to be updated, but as that's all a bit ugly, I can do it in a follow up.

Otherwise great work on this! :rocket:

**Comment by tsoj** [comment]
> Would you be willing to change this page in the docs: https://mratsim.github.io/Arraymancer/howto.perceptron.html and as source: https://github.com/mratsim/Arraymancer/blob/master/docs/howto.perceptron.rst as I'm not quite sure yet what actually changes in that example.

I don't think anything needs to be changed there (except for fixing two small bugs), as it doesn't use the nn dsl (everything is done manually).

I made a small remark to make it clear that the weird symbol->strlit->ident thingy is a workaround for the compiler issue.
